service/storage_service.cc

```diff
@@ -316,6 +316,22 @@ future<> storage_service::topology_state_load() {
         co_return *ip;
     };
 
+    if (_topology_state_machine._topology.has_complete_feature_information()) {
+        const auto enabled_features = _topology_state_machine._topology.calculate_enabled_features();
+        // We must check here that the node supports all features enabled
+        // in the cluster so far. If we won't do that then there is a risk
+        // that further commands might depend on some features that the current
+        // node doesn't understand and it might risk misinterpreting them.
+        //
+        // An interesting and useful side effect of those checks is that
+        // performing a successful read barrier makes sure that the local node
+        // supports all cluster's features.
+        const auto supported_by_me = _feature_service.supported_feature_set();
+        service::topology::check_if_enabled_features_are_known(supported_by_me, enabled_features);
```

Comment:
IIUC if the check fails, this will just crash the Raft server applier fiber and report a cryptic log message but the node will continue running. We somehow need to do the check in a controlled fashion during the startup procedure and fail the startup procedure, shutting down the node if the check fails - like in the original gossiper-based code. The check should never fail once a node has started. I'm not sure even that it belongs inside `topology_state_load()`. After all this is supposed to be a deterministic state machine so either all nodes should succeed applying a command or none should.

Reply:
I'm not sure how this check could be implemented differently in this design.

If you wanted to do it in `join_token_ring` by taking a read barrier and then checking other nodes' enabled features, then it is probably already too late - the barrier might require us to apply some commands that the node will misunderstand or not understand at all because of the missing features. In that case the group 0 state might become wrong. Something like this can happen both in the case when the node joins the cluster or is downgraded.

Reply:
Maybe we could do something like this: in `join_token_ring` when we join group 0 / do read barrier on restart, first create a channel (and destroy it after read barrier). The sending side of the channel would be in `apply`/`transfer_snapshot`. If we fail the feature check in `apply`/`transfer_snapshot` we should assert that the channel is currently initialized (we should never fail the feature check after the node has started) and push the error through the channel, so that the startup procedure can fail.

Reply:
It appears that the exception is propagated properly from `raft::server` operations after the applier fiber stopped:

```
ERROR 2023-05-16 13:51:40,699 [shard 0] init - Startup failed: raft::stopped_error (Raft instance is stopped, reason: "background error, std::_Nested_exception<raft::state_machine_error> (State machine error at raft/raft.hh:299): std::runtime_error (Feature check failed. The node doesn't support some of the features that are enabled in the cluster. Unsupported features = {TEST_ONLY_FEATURE}, supported features = {AGGREGATE_STORAGE_OPTIONS, ALTERNATOR_STREAMS, ALTERNATOR_TTL, CDC, CDC_GENERATIONS_V2, COLLECTION_INDEXING, COMPUTED_COLUMNS, CORRECT_COUNTER_ORDER, CORRECT_IDX_TOKEN_IN_SECONDARY_INDEX, CORRECT_NON_COMPOUND_RANGE_TOMBSTONES, CORRECT_STATIC_COMPACT_IN_MC, COUNTERS, DIGEST_FOR_NULL_VALUES, DIGEST_INSENSITIVE_TO_EXPIRY, DIGEST_MULTIPARTITION_READ, EMPTY_REPLICA_PAGES, HINTED_HANDOFF_SEPARATE_CONNECTION, INDEXES, KEYSPACE_STORAGE_OPTIONS, LARGE_COLLECTION_DETECTION, LARGE_PARTITIONS, LA_SSTABLE_FORMAT, LWT, MATERIALIZED_VIEWS, MC_SSTABLE_FORMAT, MD_SSTABLE_FORMAT, ME_SSTABLE_FORMAT, NONFROZEN_UDTS, PARALLELIZED_AGGREGATION, PER_TABLE_CACHING, PER_TABLE_PARTITIONERS, RANGE_SCAN_DATA_VARIANT, RANGE_TOMBSTONES, ROLES, ROW_LEVEL_REPAIR, SCHEMA_COMMITLOG, SCHEMA_TABLES_V3, SECONDARY_INDEXES_ON_STATIC_COLUMNS, SEPARATE_PAGE_SIZE_AND_SAFETY_LIMIT, STREAM_WITH_RPC_STREAM, SUPPORTS_RAFT_CLUSTER_MANAGEMENT, TOMBSTONE_GC_OPTIONS, TRUNCATION_TABLE, TYPED_ERRORS_IN_READ_RPC, UDA, UDA_NATIVE_PARALLELIZED_AGGREGATION, UDF, UNBOUNDED_RANGE_TOMBSTONES, VIEW_VIRTUAL_COLUMNS, WRITE_FAILURE_REPLY, XXHASH}, enabled features in the cluster = {AGGREGATE_STORAGE_OPTIONS, ALTERNATOR_STREAMS, ALTERNATOR_TTL, CDC, CDC_GENERATIONS_V2, COLLECTION_INDEXING, COMPUTED_COLUMNS, CORRECT_COUNTER_ORDER, CORRECT_IDX_TOKEN_IN_SECONDARY_INDEX, CORRECT_NON_COMPOUND_RANGE_TOMBSTONES, CORRECT_STATIC_COMPACT_IN_MC, COUNTERS, DIGEST_FOR_NULL_VALUES, DIGEST_INSENSITIVE_TO_EXPIRY, DIGEST_MULTIPARTITION_READ, EMPTY_REPLICA_PAGES, HINTED_HANDOFF_SEPARATE_CONNECTION, INDEXES, KEYSPACE_STORAGE_OPTIONS, LARGE_COLLECTION_DETECTION, LARGE_PARTITIONS, LA_SSTABLE_FORMAT, LWT, MATERIALIZED_VIEWS, MC_SSTABLE_FORMAT, MD_SSTABLE_FORMAT, ME_SSTABLE_FORMAT, NONFROZEN_UDTS, PARALLELIZED_AGGREGATION, PER_TABLE_CACHING, PER_TABLE_PARTITIONERS, RANGE_SCAN_DATA_VARIANT, RANGE_TOMBSTONES, ROLES, ROW_LEVEL_REPAIR, SCHEMA_COMMITLOG, SCHEMA_TABLES_V3, SECONDARY_INDEXES_ON_STATIC_COLUMNS, SEPARATE_PAGE_SIZE_AND_SAFETY_LIMIT, STREAM_WITH_RPC_STREAM, SUPPORTS_RAFT_CLUSTER_MANAGEMENT, TEST_ONLY_FEATURE, TOMBSTONE_GC_OPTIONS, TRUNCATION_TABLE, TYPED_ERRORS_IN_READ_RPC, UDA, UDA_NATIVE_PARALLELIZED_AGGREGATION, UDF, UNBOUNDED_RANGE_TOMBSTONES, VIEW_VIRTUAL_COLUMNS, WRITE_FAILURE_REPLY, XXHASH})")
```

There is a lengthy prefix indicating several layers of exception nesting, but the error message is clearly preserved. I'm not sure whether it's worth adding a channel.

> we should never fail the feature check after the node has started

This should be implicitly true - after the features of a node are persisted in the `topology` table, it is guaranteed that no features unknown to that node will be enabled (at least until the node is removed from the cluster).

Of course, if we have a bug then the above might not be true, but then we are in a pickle. If we choose to do the check anyway and stop the node if it fails, then we risk an outage but no bad things like "one node misinterprets commands from another node" will happen. On the other hand, if the node continues to work despite the failing check, there is a risk of bad things happening but we might get away with it, and we will have better availability.

I'm not sure which is better, but I think the second variant is more similar to our current behavior, so I'll go with it for the time being. Perhaps we can print an error message to logs if we detect such an issue.